Training Update

daniel112 opened the pull request:
# High level Summary
- Dadjoke generator app

Similar or identical Structure will be used on the main project

## Low level implementation
I'm focusing on architecture and structure more than the feature itself. Focusing on MVVM, Repository, and UseCase pattern

**Read the README for context and reasoning why I went with what I did.**
- [Root level README](https://github.com/daniel112/Android-Training-App/blob/3ae145ac0d0fe5b50128981ed5b8c13390a51292/README.md)
- [Dagger Hilt README](https://github.com/daniel112/Android-Training-App/blob/577f72667163322e54196fd63bf4bc6cb6aa3b3b/app/src/main/java/com/neudesic/myapplication/di/README.md)

and provide any comments/questions or suggestions on the setup :D 

## High Level diagram
- each color represents a separate file/module

![image](https://user-images.githubusercontent.com/3317835/191103324-c53bc265-1eae-4aef-9b8a-e687866e5385.png)

### Utilizes
- Dagger + Hilt Implementation for Dependency Injection
- LiveData setup for ViewModel <> View Relationship
- Retrofit plugin for HTTP calls
- basic navigation setup 

references used:
- https://blog.canopas.com/jetpack-compose-dependency-injection-with-dagger-hilt-9578be3cc1d0
- https://github.com/android/architecture-samples/tree/usecases/app/src/main/java/com/example/android/architecture/blueprints/todoapp/data/source
- https://blog.canopas.com/jetpack-compose-with-dagger-hilt-mvvm-and-navcontroller-b6048bb85073

daniel112 committed:
update:
- retrofit
- clean architecture setup
- basic navigation setup

daniel112 committed:
update:
- dagger hilt implementation
- ViewModel live data updating on HomePage

daniel112 committed:
update

ghiers18 commented on app/src/main/java/com/neudesic/myapplication/data/repository/DadJokeRepository.kt:
```diff
@@ -0,0 +1,65 @@
+class DadJokeRepositoryImpl : DadJokeRepository {
+    override suspend fun getJoke(): Response<DadJoke>? {
+        // TODO: Extract. see https://codersee.com/retrofit-with-kotlin-the-ultimate-guide/
+        val retrofit = Retrofit.Builder()
```
If you want to get real fancy, you can have dagger provide a singleton of this instance.

daniel112 replied:
I did read about that, but ive no idea how to set that up yet.
If you have a lead or pseudocode for me, that would be awesome :D

daniel112 replied:
I'd want the singleton to be able to have dynamic base URL if possible.

ghiers18 replied:
Hmm, you could have a single Retrofit instance for each api.

daniel112 replied:
done!

ghiers18 commented on app/src/main/java/com/neudesic/myapplication/data/repository/DadJokeRepository.kt:
```diff
@@ -0,0 +1,65 @@
+            .addConverterFactory(GsonConverterFactory.create())
+            .build()
+
+        val service = retrofit.create(APIService::class.java)
```
and this here would also be created via dagger

daniel112 replied:
done

ghiers18 commented on app/src/main/java/com/neudesic/myapplication/domain/models/DadJoke.kt:
```diff
@@ -0,0 +1,30 @@
+import com.google.gson.annotations.SerializedName
+
+
+data class DadJoke(
```
I recommend the kotlin serializer from retrofit. Then you can annotate the class with `@Serializable`

daniel112 replied:
is that just so I dont have to do @SerializedName on each property?

ghiers18 replied:
Yeah, it also handles the JSON differently. I believe it is _slightly faster_

daniel112 replied:
ill look to implement this soon. but slightly lower priority

ghiers18 commented on app/src/main/java/com/neudesic/myapplication/domain/repository/DadJokeRepository.kt:
```diff
@@ -0,0 +1,8 @@
+import retrofit2.Response
+
+interface DadJokeRepository {
+    suspend fun getJoke(): Response<DadJoke>?
```
I'm thinking the `Response` should not be null. Instead the `DadJoke`

daniel112 replied:
I updated it but its not optional on either. `Response<DadJoke>`
I realized that if GSON fails to convert, it will just be an empty Model of DadJoke.

I don't think its possible to get a null return type from retrofit & GSON (?)

ghiers18 replied:
That sounds right

ghiers18 commented on app/src/main/java/com/neudesic/myapplication/domain/useCase/GetDadJokesUseCase.kt:
```diff
@@ -0,0 +1,8 @@
+import com.neudesic.myapplication.domain.repository.DadJokeRepository
+import javax.inject.Inject
+
+class GetDadJokesUseCase @Inject constructor(private val dadJokeRepository: DadJokeRepository) {
```
I would suggest abstracting the "UseCase" functionality where you have an invoke or execute function. Then rename the class `GetSingleDadJokeUseCase`. So your call would look like `getSingleDadJokeUseCase.invoke()`. Reference: https://proandroiddev.com/why-you-need-use-cases-interactors-142e8a6fe576

daniel112 replied:
what if i want to support multiple cases.
like `getSingleDadJoke` and `getDadJokes`

might be a bit extra to break them down further into each own file, rather than `GetDadJokesUseCase`

I saw the article you linked, but his use case is all 1 thing.
here's another one talking about grouping it

https://proandroiddev.com/the-neverending-use-case-story-d30440e5c7f0

ghiers18 replied:
Those are separate use cases, no? You mean have several use cases (classes) in one file? In an app like this, I would say that sounds good to me.

ghiers18 commented on app/src/main/java/com/neudesic/myapplication/ui/screens/UserListScreen.kt:
```diff
@@ -0,0 +1,35 @@
+import com.neudesic.myapplication.R
+
+@Composable
+fun UserListScreen() {
```
I am useless for compose since I don't really work on UI. However, Twitter recently posted up some pretty good rules for compose development. Nonetheless, this looks good. 
https://twitter.github.io/compose-rules/rules/

daniel112 committed:
update injections

daniel112 committed:
Update:
- README updates
- cleanup

daniel112 commented:
@ghiers18 can you take a look again? Ive updated it slightly and added 2 READMEs that might help with context
- [Root level README](https://github.com/daniel112/Android-Training-App/blob/3ae145ac0d0fe5b50128981ed5b8c13390a51292/README.md)
- [Dagger Hilt README](https://github.com/daniel112/Android-Training-App/blob/bdcda8ce7c2e6fe53fea8fd90ec24064fef57538/app/src/main/java/com/neudesic/myapplication/di/README.md)

daniel112 committed:
README

daniel112 committed:
readme diagram for DI

Qwin commented on app/build.gradle:
```diff
@@ -12,6 +15,7 @@ android {
         targetSdk 32
         versionCode 1
         versionName "1.0"
+        multiDexEnabled true
```
Is multidex needed ?

daniel112 replied:
just checked. I guess it's not "required" since it's enabled by default. reference: https://developer.android.com/studio/build/multidex#:~:text=if%20your%20minSdkVersion%20is%2021%20or%20higher%20multidex%20is%20enabled%20by%20default%2C

I can remove this line.

but sounds like multDex in general is recommended to be turned on

Qwin replied:
it slows down app packaging, it used to be only required once you hit the class limit (I dont remember what it was) and multidex would allow you to split it into multiple dex files. I guess to avoid trouble in the future they set it to always enabled (although getting a compile / build time hit, but it might be not as bad nowadays)

daniel112 replied:
>  I guess to avoid trouble in the future they set it to always enabled (although getting a compile / build time hit, but it might be not as bad nowadays)

yeah, if it becomes a problem, we can specify it to be `disabled`, but I think for now going with the recommended might be a good idea

daniel112 replied:
https://github.com/daniel112/Android-Training-App/pull/1/commits/01d0ba408c4c0f310e538c3e1db82e4f6078713d

removed on this commit

Qwin commented on app/build.gradle:
```diff
@@ -30,7 +34,7 @@ android {
     }
     kotlinOptions {
-        jvmTarget = '1.8'
+        jvmTarget = '11'
```
Is there a reason to go with 11 and not a higher version ? Basically why 11 ?

daniel112 replied:
I remember Gradle 7+ requiring Java Version of at least 11. I didn't go much further than that though.

After doing a bit of research, I realize I can go higher with the gradle version (7.5.x) which supports higher Java Version
So I'll update that. As well as the Kotlin version

Here's my update for this comment:
1. updated `gradle` to be 7.5.1 (current latest)
2. Updated to the latest Android Gradle (7.2.2)
3. Updated `targetSDKVersion` and `BuildToolsVersion` to be latest (33)
4. Updated jvmTarget and `targetCompatibility` to `Java 16`  17 and 18 breaks a bunch of things. 
   - This is the latest version I could get up to without it breaking everything. 
   - error ref: https://stackoverflow.com/questions/73453524/what-is-causing-this-error-com-android-tools-r8-internal-nc-sealed-classes-are
   - https://issuetracker.google.com/issues/212279104 

ref: 
- https://docs.gradle.org/current/userguide/compatibility.html
- https://docs.gradle.org/current/userguide/compatibility.html


Let me know if I'm missing something there

daniel112 replied:
https://github.com/daniel112/Android-Training-App/pull/1/commits/01d0ba408c4c0f310e538c3e1db82e4f6078713d

Qwin replied:
I wouldnt go all the way up, my original question was more wondering why we went with java 11 vs something newer, I would check to see what is recommended nowadays for most stable version with the extra features given.

daniel112 replied:
Gotcha.
From my experience, java 11 is the most common due to latest major gradle version requiring java 11 as _minimum_.

As ive yet to see many people upgrading java version past the minimum requirements.

daniel112 replied:
buuuut, since we have it working on Java 16 right now, I'm tempted to just go with that for now until other plugins starts complaining about it 🤔

Qwin commented on app/src/main/java/com/neudesic/myapplication/ui/screen/home/HomeViewModel.kt:
```diff
@@ -0,0 +1,39 @@
+@HiltViewModel
+class HomeViewModel @Inject constructor(private val getDadJokesUseCase: GetDadJokesUseCase): ViewModel() {
+    private val _dadJoke = MutableLiveData<DadJoke>(null)
+    val dadJoke: LiveData<DadJoke> = _dadJoke
```
So even though DadJoke is the model (DTO), we'd want to make a DadJokeChildViewModel, which then represents what the view represents, this will abstract what data is coming in from the service to what we want to display.

Qwin commented:
Another suggestion is to add another domain which has a different screen, where the navigation is used to go to that page and pass the right viewModel, with its own stack of useCase + repo, why?  because then a person that needs to add a new feature / screen would know exactly how to it. It also gives a good perspective if the navigation is working as expected.

Lastly add some docs for setup and how to work with Dagger+Hilt.

daniel112 committed:
update:
- remove multiDexEnabled since its on by default
- updated to latest possible gradle version (7.5.1)

daniel112 committed:
update:
- abstraction with DTO > Domain model
- added retrofit adapterFactory to abstract away the boilerplate of converting Retrofit Response to our custom DataResult

daniel112 commented:
merging this so I can keep better track of progress
